File: java/egy-names/src/test/java/com/afify/egynames/EgyptianNamesTest.java
package com.afify.egynames;

import com.afify.egynames.model.Models;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;

import java.util.List;
import java.util.Map;

import static org.junit.jupiter.api.Assertions.*;

public class EgyptianNamesTest {

    private static EgyptianNames en;

    @BeforeAll
    public static void setUp() {
        en = new EgyptianNames();
    }

    @Test
    public void testStats() {
        Map<String, Object> stats = en.stats();
        assertNotNull(stats);
        int total = (int) stats.get("total_names");
        assertTrue(total > 30000, "Loaded names should exceed 30,000");
    }

    @Test
    public void testTashkeelAnd11D() {
        String tk = en.tashkeel("محمد عبدالرحمن");
        assertTrue(tk.contains("مُحَمَّد"));
        assertTrue(tk.contains("الرَّحْمَن"));

        String tkEg = en.tashkeelEg("محمد");
        assertFalse(tkEg.isEmpty());

        String ipaStd = en.ipa("جمال", "standard");
        assertTrue(ipaStd.startsWith("/"));

        String ipaEg = en.ipaEg("جمال");
        assertTrue(ipaEg.startsWith("["));

        List<String> dallaa = en.dallaa("محمد");
        assertTrue(dallaa.contains("ميدو"));

        String root = en.root("محمد");
        assertEquals("ح-م-د", root);

        String origin = en.origin("محمد");
        assertEquals("arabic_classical", origin);

        List<String> figures = en.famousFigures("محمد");
        assertFalse(figures.isEmpty());

        String trend = en.trend("محمد");
        assertEquals("classic_timeless", trend);
    }

    @Test
    public void testGenerationAndDemographics() {
        List<Models.GeneratedName> names = en.generate(3, "male", "muslim");
        assertEquals(3, names.size());

        Models.GenderDetection g = en.detectGender("مريم إبراهيم حسن");
        assertEquals("female", g.gender);

        Models.ReligionDetection r = en.detectReligion("جورج بطرس سمير ميخائيل");
        assertEquals("christian", r.religion);
    }
}

File: dart/egy_names/test/egyptian_names_test.dart
import 'package:egy_names/egy_names.dart';
import 'package:test/test.dart';

void main() {
  group('EgyNames Dart Tests', () {
    final en = EgyNames(seed: 42);

    test('Data statistics', () {
      final stats = en.stats();
      expect(stats['total_names'], greaterThan(30000));
      expect(stats['corpus_students'], greaterThan(2000000));
    });

    test('Generation works', () {
      final names = en.generate(count: 5);
      expect(names.length, equals(5));
      for (final n in names) {
        expect(n.ar.isNotEmpty, isTrue);
        expect(n.en.isNotEmpty, isTrue);
      }
    });

    test('Translation works in both directions', () {
      expect(en.translate('محمد أحمد علي'), equals('Mohamed Ahmed Ali'));
      expect(en.translate('Mohamed Ahmed Ali'), equals('محمد أحمد علي'));
    });

    test('Splitting DP segmentation', () {
      final splitResult = en.split('محمدأحمدعليحسنالشاذلي');
      expect(splitResult, equals(['محمد', 'أحمد', 'علي', 'حسن', 'الشاذلي']));
    });

    test('Correction & Tashkeel', () {
      expect(en.correct('احمد'), equals('أحمد'));
      expect(en.tashkeel('محمد عبدالرحمن'), contains('مُحَمَّد'));
      expect(en.tashkeel('محمد عبدالرحمن'), contains('الرَّحْمَن'));
    });

    test('Meaning annotation', () {
      final meaning = en.meaning('محمد');
      expect(meaning, isNotNull);
      expect(meaning!['ar'], contains('المحمود'));
    });

    test('11D Features (Tashkeel Eg, IPA, Dallaa, Roots, Trends)', () {
      expect(en.tashkeelEg('محمد'), isNotEmpty);
      expect(en.ipa('جمال', dialect: 'standard'), startsWith('/'));
      expect(en.ipaEg('جمال'), startsWith('['));
      expect(en.dallaa('محمد'), contains('ميدو'));
      expect(en.root('محمد'), equals('ح-م-د'));
      expect(en.origin('محمد'), equals('arabic_classical'));
      expect(en.famousFigures('محمد'), isNotEmpty);
      expect(en.trend('محمد'), equals('classic_timeless'));
    });

    test('Creative gender and religion detection', () {
      final g = en.detectGender('مريم إبراهيم حسن');
      expect(g.gender, equals('female'));

      final r = en.detectReligion('جورج بطرس سمير ميخائيل');
      expect(r.religion, equals('christian'));
    });

    test('First given name wins; non-person surfaces are not valid', () {
      expect(en.detectGender('فاطمة محمد علي حسن').gender, equals('female'));
      expect(
        en.detectReligion('جورج علاءالدين عبدالمسيح دغيدي').religion,
        equals('christian'),
      );
      expect(en.isValid('الله'), isFalse);
      expect(en.translate('Mahmoud'), equals('محمود'));
    });
  });
}
